// @pnp/graph-commonjs/calendars/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Events = exports._Events = exports.Event = exports._Event = exports.Calendars = exports._Calendars = exports.Calendar = exports._Calendar = void 0;
var tslib_1 = require("tslib");
var odata_1 = require("@pnp/odata-commonjs");
var graphqueryable_js_1 = require("../graphqueryable.js");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var funcs_js_1 = require("./funcs.js");
/**
 * Calendar
 */
var _Calendar = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Calendar, _super);
    function _Calendar() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.calendarView = funcs_js_1.calendarView;
        return _this;
    }
    Object.defineProperty(_Calendar.prototype, "events", {
        get: function () {
            return (0, exports.Events)(this);
        },
        enumerable: false,
        configurable: true
    });
    return _Calendar;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._Calendar = _Calendar;
exports.Calendar = (0, graphqueryable_js_1.graphInvokableFactory)(_Calendar);
/**
 * Calendars
 */
var _Calendars = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Calendars, _super);
    function _Calendars() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _Calendars = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("calendars"),
        (0, decorators_js_1.getById)(exports.Calendar)
    ], _Calendars);
    return _Calendars;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._Calendars = _Calendars;
exports.Calendars = (0, graphqueryable_js_1.graphInvokableFactory)(_Calendars);
/**
 * Event
 */
var _Event = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Event, _super);
    function _Event() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.instances = funcs_js_1.instances;
        return _this;
    }
    _Event = (0, tslib_1.__decorate)([
        (0, decorators_js_1.deleteable)(),
        (0, decorators_js_1.updateable)()
    ], _Event);
    return _Event;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._Event = _Event;
exports.Event = (0, graphqueryable_js_1.graphInvokableFactory)(_Event);
/**
 * Events
 */
var _Events = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Events, _super);
    function _Events() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Adds a new event to the collection
     *
     * @param properties The set of properties used to create the event
     */
    _Events.prototype.add = function (properties) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.graphPost)(this, (0, odata_1.body)(properties))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                event: this.getById(data.id),
                            }];
                }
            });
        });
    };
    _Events = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("events"),
        (0, decorators_js_1.getById)(exports.Event)
    ], _Events);
    return _Events;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._Events = _Events;
exports.Events = (0, graphqueryable_js_1.graphInvokableFactory)(_Events);
