// @pnp/graph-commonjs/insights/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Resource = exports._Resource = exports.SharedInsights = exports._SharedInsights = exports.SharedInsight = exports._SharedInsight = exports.UsedInsights = exports._UsedInsights = exports.UsedInsight = exports._UsedInsight = exports.TrendingInsights = exports._TrendingInsights = exports.TrendingInsight = exports._TrendingInsight = exports.Insights = exports._Insights = void 0;
var tslib_1 = require("tslib");
var graphqueryable_js_1 = require("../graphqueryable.js");
var decorators_js_1 = require("../decorators.js");
/**
 * Represents a insights entity
 */
var _Insights = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Insights, _super);
    function _Insights() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_Insights.prototype, "trending", {
        get: function () {
            return (0, exports.TrendingInsights)(this);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Insights.prototype, "used", {
        get: function () {
            return (0, exports.UsedInsights)(this);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Insights.prototype, "shared", {
        get: function () {
            return (0, exports.SharedInsights)(this);
        },
        enumerable: false,
        configurable: true
    });
    _Insights = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("insights")
    ], _Insights);
    return _Insights;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._Insights = _Insights;
exports.Insights = (0, graphqueryable_js_1.graphInvokableFactory)(_Insights);
/**
 * Describes a Trending Insight instance
 */
var _TrendingInsight = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_TrendingInsight, _super);
    function _TrendingInsight() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_TrendingInsight.prototype, "resource", {
        get: function () {
            return (0, exports.Resource)(this);
        },
        enumerable: false,
        configurable: true
    });
    return _TrendingInsight;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._TrendingInsight = _TrendingInsight;
exports.TrendingInsight = (0, graphqueryable_js_1.graphInvokableFactory)(_TrendingInsight);
/**
 * Describes a collection of Trending Insight objects
 *
 */
var _TrendingInsights = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_TrendingInsights, _super);
    function _TrendingInsights() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _TrendingInsights = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("trending"),
        (0, decorators_js_1.getById)(exports.TrendingInsight)
    ], _TrendingInsights);
    return _TrendingInsights;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._TrendingInsights = _TrendingInsights;
exports.TrendingInsights = (0, graphqueryable_js_1.graphInvokableFactory)(_TrendingInsights);
/**
 * Describes a Used Insight instance
 */
var _UsedInsight = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_UsedInsight, _super);
    function _UsedInsight() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_UsedInsight.prototype, "resource", {
        get: function () {
            return (0, exports.Resource)(this);
        },
        enumerable: false,
        configurable: true
    });
    return _UsedInsight;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._UsedInsight = _UsedInsight;
exports.UsedInsight = (0, graphqueryable_js_1.graphInvokableFactory)(_UsedInsight);
/**
 * Describes a collection of Used Insight objects
 *
 */
var _UsedInsights = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_UsedInsights, _super);
    function _UsedInsights() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _UsedInsights = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("used"),
        (0, decorators_js_1.getById)(exports.UsedInsight)
    ], _UsedInsights);
    return _UsedInsights;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._UsedInsights = _UsedInsights;
exports.UsedInsights = (0, graphqueryable_js_1.graphInvokableFactory)(_UsedInsights);
/**
 * Describes a Shared Insight instance
 */
var _SharedInsight = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_SharedInsight, _super);
    function _SharedInsight() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_SharedInsight.prototype, "resource", {
        get: function () {
            return (0, exports.Resource)(this);
        },
        enumerable: false,
        configurable: true
    });
    return _SharedInsight;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._SharedInsight = _SharedInsight;
exports.SharedInsight = (0, graphqueryable_js_1.graphInvokableFactory)(_SharedInsight);
/**
 * Describes a collection of Shared Insight objects
 *
 */
var _SharedInsights = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_SharedInsights, _super);
    function _SharedInsights() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _SharedInsights = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("shared"),
        (0, decorators_js_1.getById)(exports.SharedInsight)
    ], _SharedInsights);
    return _SharedInsights;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._SharedInsights = _SharedInsights;
exports.SharedInsights = (0, graphqueryable_js_1.graphInvokableFactory)(_SharedInsights);
/**
 * Describes a Resource Entity instance
 */
var _Resource = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Resource, _super);
    function _Resource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _Resource = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("resource")
    ], _Resource);
    return _Resource;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._Resource = _Resource;
exports.Resource = (0, graphqueryable_js_1.graphInvokableFactory)(_Resource);
