// @pnp/graph-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toAbsoluteUrl = exports.GraphQueryableSearchableCollection = exports.GraphQueryableInstance = exports.GraphQueryableCollection = exports.GraphQueryable = exports.GraphBatch = exports.GraphRest = exports.graph = void 0;
var tslib_1 = require("tslib");
var rest_js_1 = require("./rest.js");
Object.defineProperty(exports, "graph", { enumerable: true, get: function () { return rest_js_1.graph; } });
Object.defineProperty(exports, "GraphRest", { enumerable: true, get: function () { return rest_js_1.GraphRest; } });
var batch_js_1 = require("./batch.js");
Object.defineProperty(exports, "GraphBatch", { enumerable: true, get: function () { return batch_js_1.GraphBatch; } });
var graphqueryable_js_1 = require("./graphqueryable.js");
Object.defineProperty(exports, "GraphQueryable", { enumerable: true, get: function () { return graphqueryable_js_1.GraphQueryable; } });
Object.defineProperty(exports, "GraphQueryableCollection", { enumerable: true, get: function () { return graphqueryable_js_1.GraphQueryableCollection; } });
Object.defineProperty(exports, "GraphQueryableInstance", { enumerable: true, get: function () { return graphqueryable_js_1.GraphQueryableInstance; } });
Object.defineProperty(exports, "GraphQueryableSearchableCollection", { enumerable: true, get: function () { return graphqueryable_js_1.GraphQueryableSearchableCollection; } });
(0, tslib_1.__exportStar)(require("./graphhttpclient.js"), exports);
(0, tslib_1.__exportStar)(require("./types.js"), exports);
(0, tslib_1.__exportStar)(require("./operations.js"), exports);
var toabsoluteurl_js_1 = require("./utils/toabsoluteurl.js");
Object.defineProperty(exports, "toAbsoluteUrl", { enumerable: true, get: function () { return toabsoluteurl_js_1.toAbsoluteUrl; } });
