// @pnp/graph-commonjs/rest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.graph = exports.GraphRest = void 0;
var tslib_1 = require("tslib");
var graphlibconfig_js_1 = require("./graphlibconfig.js");
var batch_js_1 = require("./batch.js");
var common_1 = require("@pnp/common-commonjs");
var GraphRest = /** @class */ (function () {
    /**
     * Creates a new instance of the SPRest class
     *
     * @param options Additional options
     * @param baseUrl A string that should form the base part of the url
     */
    function GraphRest(_options, _baseUrl, _runtime) {
        if (_options === void 0) { _options = {}; }
        if (_baseUrl === void 0) { _baseUrl = "v1.0"; }
        if (_runtime === void 0) { _runtime = common_1.DefaultRuntime; }
        this._options = _options;
        this._baseUrl = _baseUrl;
        this._runtime = _runtime;
    }
    GraphRest.prototype.createBatch = function () {
        return new batch_js_1.GraphBatch().setRuntime(this._runtime);
    };
    GraphRest.prototype.setup = function (config) {
        if (config.pageContext) {
            (0, graphlibconfig_js_1.setup)({
                spfxContext: config,
            }, this._runtime);
        }
        else {
            (0, graphlibconfig_js_1.setup)(config, this._runtime);
        }
    };
    GraphRest.prototype.createIsolated = function (init) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var baseUrl, cloneGlobal, options, config, runtime;
            return (0, tslib_1.__generator)(this, function (_a) {
                // merge our defaults
                init = Object.assign({
                    baseUrl: "v1.0",
                    cloneGlobal: true,
                    config: {},
                    options: {},
                }, init || {});
                baseUrl = init.baseUrl, cloneGlobal = init.cloneGlobal, options = init.options, config = init.config;
                runtime = cloneGlobal ? new common_1.Runtime(common_1.DefaultRuntime.export()) : new common_1.Runtime();
                runtime.assign(config);
                return [2 /*return*/, new GraphRest(options, baseUrl, runtime)];
            });
        });
    };
    GraphRest.prototype.childConfigHook = function (callback) {
        return callback({ options: this._options, baseUrl: this._baseUrl, runtime: this._runtime });
    };
    return GraphRest;
}());
exports.GraphRest = GraphRest;
exports.graph = new GraphRest();
