// @pnp/graph-commonjs/onenote/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sections = exports._Sections = exports.Section = exports._Section = exports.Notebooks = exports._Notebooks = exports.Notebook = exports._Notebook = exports.OneNote = exports._OneNote = void 0;
var tslib_1 = require("tslib");
var odata_1 = require("@pnp/odata-commonjs");
var graphqueryable_js_1 = require("../graphqueryable.js");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
/**
 * Represents a onenote entity
 */
var _OneNote = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_OneNote, _super);
    function _OneNote() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_OneNote.prototype, "notebooks", {
        get: function () {
            return (0, exports.Notebooks)(this);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_OneNote.prototype, "sections", {
        get: function () {
            return (0, exports.Sections)(this);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_OneNote.prototype, "pages", {
        get: function () {
            return this.clone(graphqueryable_js_1.GraphQueryableCollection, "pages");
        },
        enumerable: false,
        configurable: true
    });
    _OneNote = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("onenote")
    ], _OneNote);
    return _OneNote;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._OneNote = _OneNote;
exports.OneNote = (0, graphqueryable_js_1.graphInvokableFactory)(_OneNote);
/**
 * Describes a notebook instance
 *
 */
var _Notebook = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Notebook, _super);
    function _Notebook() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_Notebook.prototype, "sections", {
        get: function () {
            return (0, exports.Sections)(this);
        },
        enumerable: false,
        configurable: true
    });
    return _Notebook;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._Notebook = _Notebook;
exports.Notebook = (0, graphqueryable_js_1.graphInvokableFactory)(_Notebook);
/**
 * Describes a collection of Notebook objects
 *
 */
var _Notebooks = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Notebooks, _super);
    function _Notebooks() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a new notebook as specified in the request body.
     *
     * @param displayName Notebook display name
     */
    _Notebooks.prototype.add = function (displayName) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.graphPost)(this, (0, odata_1.body)({ displayName: displayName }))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                notebook: this.getById(data.id),
                            }];
                }
            });
        });
    };
    _Notebooks = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("notebooks"),
        (0, decorators_js_1.getById)(exports.Notebook)
    ], _Notebooks);
    return _Notebooks;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._Notebooks = _Notebooks;
exports.Notebooks = (0, graphqueryable_js_1.graphInvokableFactory)(_Notebooks);
/**
 * Describes a sections instance
 */
var _Section = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Section, _super);
    function _Section() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return _Section;
}(graphqueryable_js_1._GraphQueryableInstance));
exports._Section = _Section;
exports.Section = (0, graphqueryable_js_1.graphInvokableFactory)(_Section);
/**
 * Describes a collection of Sections objects
 *
 */
var _Sections = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Sections, _super);
    function _Sections() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Adds a new section
     *
     * @param displayName New section display name
     */
    _Sections.prototype.add = function (displayName) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.graphPost)(this, (0, odata_1.body)({ displayName: displayName }))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                section: this.getById(data.id),
                            }];
                }
            });
        });
    };
    _Sections = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("sections"),
        (0, decorators_js_1.getById)(exports.Section)
    ], _Sections);
    return _Sections;
}(graphqueryable_js_1._GraphQueryableCollection));
exports._Sections = _Sections;
exports.Sections = (0, graphqueryable_js_1.graphInvokableFactory)(_Sections);
