// @pnp/graph-commonjs/teams/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Teams = exports.Team = exports.Tabs = exports.Tab = exports.Channels = exports.Channel = void 0;
var tslib_1 = require("tslib");
var odata_1 = require("@pnp/odata-commonjs");
var rest_1 = require("../rest");
var types_1 = require("../groups/types");
var types_2 = require("./types");
var operations_1 = require("../operations");
require("./users");
var types_js_1 = require("./types.js");
Object.defineProperty(exports, "Channel", { enumerable: true, get: function () { return types_js_1.Channel; } });
Object.defineProperty(exports, "Channels", { enumerable: true, get: function () { return types_js_1.Channels; } });
Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return types_js_1.Tab; } });
Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return types_js_1.Tabs; } });
Object.defineProperty(exports, "Team", { enumerable: true, get: function () { return types_js_1.Team; } });
Object.defineProperty(exports, "Teams", { enumerable: true, get: function () { return types_js_1.Teams; } });
(0, odata_1.addProp)(types_1._Group, "team", types_2.Team);
types_1._Group.prototype.createTeam = function (props) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var data;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, operations_1.graphPut)(this.clone(types_1.Group, "team"), (0, odata_1.body)(props))];
                case 1:
                    data = _a.sent();
                    return [2 /*return*/, {
                            data: data,
                            team: this.team,
                        }];
            }
        });
    });
};
Reflect.defineProperty(rest_1.GraphRest.prototype, "teams", {
    configurable: true,
    enumerable: true,
    get: function () {
        return this.childConfigHook(function (_a) {
            var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
            return (0, types_2.Teams)(baseUrl).configure(options).setRuntime(runtime);
        });
    },
});
