// @pnp/graph-commonjs/planner/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlanDetails = exports.TaskDetails = exports.Tasks = exports.Task = exports.Plans = exports.Planner = exports.Plan = exports.Buckets = exports.Bucket = void 0;
var rest_js_1 = require("../rest.js");
var types_js_1 = require("./types.js");
require("./groups.js");
require("./users.js");
var types_js_2 = require("./types.js");
Object.defineProperty(exports, "Bucket", { enumerable: true, get: function () { return types_js_2.Bucket; } });
Object.defineProperty(exports, "Buckets", { enumerable: true, get: function () { return types_js_2.Buckets; } });
Object.defineProperty(exports, "Plan", { enumerable: true, get: function () { return types_js_2.Plan; } });
Object.defineProperty(exports, "Planner", { enumerable: true, get: function () { return types_js_2.Planner; } });
Object.defineProperty(exports, "Plans", { enumerable: true, get: function () { return types_js_2.Plans; } });
Object.defineProperty(exports, "Task", { enumerable: true, get: function () { return types_js_2.Task; } });
Object.defineProperty(exports, "Tasks", { enumerable: true, get: function () { return types_js_2.Tasks; } });
Object.defineProperty(exports, "TaskDetails", { enumerable: true, get: function () { return types_js_2.TaskDetails; } });
Object.defineProperty(exports, "PlanDetails", { enumerable: true, get: function () { return types_js_2.PlanDetails; } });
Reflect.defineProperty(rest_js_1.GraphRest.prototype, "planner", {
    configurable: true,
    enumerable: true,
    get: function () {
        return this.childConfigHook(function (_a) {
            var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
            return (0, types_js_1.Planner)(baseUrl).configure(options).setRuntime(runtime);
        });
    },
});
